pkg/receive/hashring.go

```diff
@@ -261,41 +260,103 @@ func newMultiHashring(algorithm HashringAlgorithm, replicationFactor uint64, cfg
 	return m
 }
 
+type HashringUpdate struct {
+	PreviousConfig []HashringConfig
+	NewConfig      []HashringConfig
+	Hashring       Hashring
+}
+
+// UpdatedTenants returns all tenants whose hashrings have been updated.
+// Tenants whose hashrings do not include ownEndpoint are not considered
+// as updated. This allows for minimizing disruptions when flushing the
+// TSDB head when hashrings change.
+func (h HashringUpdate) UpdatedTenants(ownEndpoint string) []string {
+	newHashrings := hashringsByTenant(h.NewConfig)
+	oldHashrings := hashringsByTenant(h.PreviousConfig)
+
+	var updates []string
+	// Include tenants that were removed and the endpoint was a part of their hashring.
+	for t, h := range oldHashrings {
+		hasEndpoint := slices.Contains(h.Endpoints, ownEndpoint)
+		if _, kept := newHashrings[t]; !kept && hasEndpoint {
+			updates = append(updates, t)
+		}
+	}
+
+	for t, newHashring := range newHashrings {
+		oldHashring, ok := oldHashrings[t]
+		if !ok {
+			continue
+		}
+
+		endpointInHashring := slices.Contains(oldHashring.Endpoints, ownEndpoint) || slices.Contains(newHashring.Endpoints, ownEndpoint)
+		if !endpointInHashring {
+			continue
+		}
+
+		if !slices.Equal(oldHashring.Endpoints, newHashring.Endpoints) {
+			updates = append(updates, t)
+		}
+	}
+	return updates
+}
+
+func hashringsByTenant(configs []HashringConfig) map[string]HashringConfig {
+	byTenant := make(map[string]HashringConfig)
+	for _, newHashring := range configs {
+		for _, t := range newHashring.Tenants {
```

Comment:
I think I'm also a bit confused by the different variants and when we do (not) need to flush TSDB. So maybe it would help us to break it down? My understanding now is similar to what @saswatamcode suggests:
- If we change `endpoints` in a hashring - we want to always flush all TSDBs in that hashring
- If we do **not** change `endpoints`, but add new entries in `tenants` - no need to flush TSDBs in that hashring
- If we do **not** change `endpoints`, but remove one or more `tenants` - we want to flush TSDBs of tenants that were removed

Is this correct?